jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannel.java

```diff
@@ -493,34 +454,73 @@ public void onFrame(Frame frame, Callback callback)
     @Override
     public void sendFrame(Frame frame, Callback callback, boolean batch)
     {
-        if (LOG.isDebugEnabled())
-            LOG.debug("sendFrame({}, {}, {})", frame, callback, batch);
-
-        try
-        {
-            assertValidOutgoing(frame);
-            outgoingSequence.check(frame.getOpCode(), frame.isFin());
-        }
-        catch (Throwable ex)
+        synchronized(this)
```

Comment:
I fail to see the need for this `synchronized`.

Inside its block, it only works on local variables (so no need for sync) and at the end it enqueues the frame into the `FrameFlusher`, which is already synchronized.
And it then calls `connection.sendFrameQueue()`, which is just `FrameFlusher.iterate()`, outside of the sync block.

Reply:
The main purpose is to allow only a single frame into the extension stack at once and to avoid parallel sends overtaking each other in the extension stack.   Once the `checkoutOutgoing` is done, we must exclude other threads until the frame is queued - otherwise the ordering that is  checked may become invalid.

Reply:
I don't see how the synchronized will help.

If the upper layer is entitled to send frames, then 2 concurrent writes can interleave their frames, resulting in garbled bytes.

So if it is responsibility of the upper layer to avoid this, then the lower layer API needs no synchronization - just documentation about how to use it.

Reply:
The upper layer should avoid this, but it is still the cores responsibility to ensure that bad frame sequences are not transmitted on the wire, so it cannot allow funky multi threaded calls to get past its sequence checking... even if it is a bad application.  

So as discussed in hangout review, we leave synchronized for now